File: src/main/java/com/myapp/auth/StatelessAuthenticationFilter.java
package com.myapp.auth;

import io.jsonwebtoken.JwtException;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.GenericFilterBean;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.ServletRequest;
import javax.servlet.ServletResponse;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

@Component
class StatelessAuthenticationFilter extends GenericFilterBean {

    private final TokenAuthenticationService tokenAuthenticationService;

    StatelessAuthenticationFilter(TokenAuthenticationService tokenAuthenticationService) {
        this.tokenAuthenticationService = tokenAuthenticationService;
    }

    @Override
    public void doFilter(ServletRequest req, ServletResponse res, FilterChain chain) throws IOException, ServletException {
        try {
            Authentication authentication = tokenAuthenticationService.getAuthentication((HttpServletRequest) req);
            SecurityContextHolder.getContext().setAuthentication(authentication);
            chain.doFilter(req, res);
            SecurityContextHolder.getContext().setAuthentication(null);
        } catch (AuthenticationException | JwtException e) {
            SecurityContextHolder.clearContext();
            ((HttpServletResponse) res).setStatus(HttpServletResponse.SC_UNAUTHORIZED);
        }
    }
}

File: src/main/kotlin/com/myapp/auth/StatelessAuthenticationFilter.kt
package com.myapp.auth

import io.jsonwebtoken.JwtException
import org.springframework.security.core.AuthenticationException
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Component
import org.springframework.web.filter.GenericFilterBean
import javax.servlet.FilterChain
import javax.servlet.ServletRequest
import javax.servlet.ServletResponse
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@Component
class StatelessAuthenticationFilter(
    private val tokenAuthenticationService: TokenAuthenticationService
) : GenericFilterBean() {

    override fun doFilter(req: ServletRequest, res: ServletResponse, chain: FilterChain) {
        try {
            val authentication = tokenAuthenticationService.authentication(req as HttpServletRequest)
            SecurityContextHolder.getContext().authentication = authentication
            chain.doFilter(req, res)
            SecurityContextHolder.getContext().authentication = null
        } catch (e: AuthenticationException) {
            SecurityContextHolder.clearContext()
            (res as HttpServletResponse).status = HttpServletResponse.SC_UNAUTHORIZED
        } catch (e: JwtException) {
            SecurityContextHolder.clearContext()
            (res as HttpServletResponse).status = HttpServletResponse.SC_UNAUTHORIZED
        }
    }

}
